### PR TITLE
Include parens to type parameter

### PR DESCRIPTION
**Pull request by topecongiro**
The motivation of this PR is to fix a bug in rustfmt (cc https://github.com/rust-lang-nursery/rustfmt/issues/2630).

### CODE REVIEW

**Comment by estebank** [comment]
Can you add a test exercising the new code? Other than that, LGTM.

**Comment by topecongiro** [comment]
@estebank Thank you for the feedback! Though I am not sure how I should test the new code, I will try writing some ui tests and see if they work...

File: src/test/ui/maybe-bounds.stderr [hunk 1]
```diff
@@ -1,22 +1,22 @@
 error: `?Trait` is not permitted in supertraits
-  --> $DIR/maybe-bounds.rs:11:12
+  --> $DIR/maybe-bounds.rs:11:11
    |
 LL | trait Tr: ?Sized {} //~ ERROR `?Trait` is not permitted in supertraits
-   |            ^^^^^
```
Comment (petrochenkov) [inline src/test/ui/maybe-bounds.stderr]: @topecongiro 
>Though I am not sure how I should test the new code

You can put any of `?Sized`s in this test into parentheses, that would be enough.

File: src/libsyntax/parse/parser.rs [hunk 2]
```diff
@@ -4671,7 +4677,6 @@ impl<'a> Parser<'a> {
                     bounds.push(TraitTyParamBound(poly_trait, modifier));
                 }
                 if has_parens {
-                    self.expect(&token::CloseDelim(token::Paren))?;
                     if let Some(&RegionTyParamBound(..)) = bounds.last() {
                         self.span_err(self.prev_span,
                                       "parenthesized lifetime bounds are not supported");
```
Comment (petrochenkov) [inline src/libsyntax/parse/parser.rs]: Nit: Could you move this error into `if self.token.is_lifetime()` above as well?

**Comment by topecongiro** [comment]
@petrochenkov Thank you for the review! I am going to work on this over the weekend.

**Comment by pietroalbini** [comment]
Ping from triage @topecongiro! It's been a while since we heard from you, will you have time to work on this again soon?

**Comment by topecongiro** [comment]
My apologies for the late response. Hopefully over this weekend I may have some time...

**Comment by shepmaster** [comment]
Ping from triage, @topecongiro! How's your weekend looking this time around?

**Comment by topecongiro** [comment]
I have updated the tests and fixed the code based on the review. I am very sorry that this took so long...

**Comment by pietroalbini** [comment]
@topecongiro could you please rebase your branch into master (even if it doesn't show any merge conflict)? Thanks!

**Comment by topecongiro** [comment]
Rebased.

**Comment by petrochenkov** [comment]
Network failure.
@bors retry

**Pull request merged**